rfcs/rfc-core-2023-04-20-new-archive-tdd.md

```diff
@@ -0,0 +1,1559 @@
+
+*   The ability to Create/Remove Directories within the Archive
+*   Querying FilesystemAttributes (type\[directory, file, symlink\], modtime, permissions,etc...) within the archive
+*   Search for files via globbing directly as part of the archive API  
+    This can be done at a higher level with another entry in the Streamer Stack, that builds a directory tree out of the mounted archive files  
+    
+*   Resolving FileIOAliases  
+    i.e @product@/level.archive/asset.txt, → D:/o3de/AutomatedTesting/level.archive/asset.txt
+
+The CryPak archive API can create and query ZIP Format, with support for 3 hard-coded Compression Codecs.  
+The use of the Compression Algorithms of Zlib, LZ4, ZSTD, will NOT be implemented as part of the new Archive API system itself,  
+but instead layered as an entry in the Streamer that can perform decompression after streamer has read the data from an Archive file
+
+Searching for Files within the multiple mounted Archive will be managed by the Archive Streamer stack node, that will provide a mechanism for prioriting file lookups within the archive itself.  
+The search priority between loose files and archive files, will be completely kept out of the Archive System, and will take advantage of the how nodes are order in Streamer stack.  
+As the Streamer stack is configurable through the Settings Registry, this is user controllable.  
+
+New Features
+-----------------
+
+The following features would be added for the archive format
+
+### Format Features
+
+1.  Support for Archive files over 4GiB  
+    The limit for single archive file would be up to 170TiB, with support for up to 33 million files.  
+    CryPak currently only supports the standard ZIP Format and not ZIP64, despite the [ZipFileFormat.h](https://github.com/o3de/o3de/blob/5f5dddf46a8a2b6609c2088e70793e89f5ff48dd/Code/Framework/AzFramework/AzFramework/Archive/ZipFileFormat.h#L129-L141) header implying such support.  
+    Only 32-bit compressed and uncompressed sizes are [read](https://github.com/o3de/o3de/blob/5f5dddf46a8a2b6609c2088e70793e89f5ff48dd/Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp#L107) from a compressed entry with no support for 64-bit values.  
+2.  Archive header stored at the beginning of the Archive  
+    This allows mounting the archive to only read the uncompressed section of the archive header(which is 80 bytes currently padded up to 512-bytes) for future entries.  
+    This can then be used to located the Table of Contents at the end of the Archive in a single seek.
+3.  The Archive Table of Contents (TOC) is stored at the end of the Archive file  
+    There is a TOC at the end of the Archive File that is variable length.  
+    The TOC is optionally compressed which will be indicated within the Archive header  
+    The Archive format supports up to 2^25 or ~33 million files(This is a soft cap in the format, technically up to 2^32 can be stored, but it will be restricted to 2^25 for now)  
```

Comment:
why the soft cap?

Reply:
This is because the block offset table will only supports 25-bits of for indices and therefore only support ~33 million blocks.

Because only 2^25 blocks of data can be represented in the archive, the maximum number of files is restricted to 2^25 as well.